### ISSUE TITLE
Papercuts

### ISSUE THREAD

**Issue by tannerhelland** [tannerhelland/PhotoDemon#164]
I love Ubuntu's notion of "papercuts" (https://launchpad.net/hundredpapercuts) -- small annoyances that are not particularly difficult to fix, but that negatively impact UX.

PD is (unfortunately) rife with papercuts, as I tend to spend a lot of time adding new features instead of perfecting existing ones.  To that end, I'd like to start publicly tracking papercuts, and spend at least a little time each week dealing with them.

This list is not exhaustive.  I write things here when I think of them, but don't have time to address them immediately.  Outside recommendations are of course welcome; I'll update this list with whatever comments I receive from users and/or developers.
  - [x] ~Individual tool presets can't be edited or removed once added.~  *As of a6c8ebef78f5805b8f35a2752c78a4b2a90d7d63, this feature is now available*
  - [x] Tab key navigation is not well tested, in part because many PD controls do not render themselves differently upon receiving keyboard focus.  *Many improvements are now available in nightly builds; latest work includes 8e9b582c9223de865742b542aa7703b3e8fdfb0a*

*updated 21 March 2018*

**Comment by jpbro** [comment]
Regarding the tab order "mess". One thing I can do is run through all of the forms and click use the automatic tab ordering of MZ Tools if you'd like. It won't be perfect, but I suspect it will improve things a bit.

One issue may be the pdFxPreviewControls - they will usually end up as the first tab item, and I'm not sure this is what you would want. I can manually move them down the tab stack but I'd like to know where you'd like them to end up - after the bottom cmdBar control, immediately before the bottom cmdBar control, or somewhere else?

Also can you think of any other common controls that might be automatically moved to the wrong tab order that I would need to manually adjust?

**Comment by tannerhelland** [comment]
Thank you for taking a look at this!  I completely forget to update these older issues as I go.  *(insert sheepish look here)*

In 7.0, I wrote new functions to automatically assigns tab order at run-time.  The primary commit was this one, I believe:

https://github.com/tannerhelland/PhotoDemon/commit/75df46048e1f94ddd42e38b0bfaa1da665389064

That went a long way toward solving tab order problems, which have plagued me since Day 1 because I'm constantly messing with dialog layouts.  

I am worried, however, that there may be bugs similar to the ones you've turned up with Enter/Esc key functionality.  As I note in that commit description, the feature is also hindered because many of my custom controls don't redraw themselves when they receive keyboard focus.  (Instead, all their UI changes are based around mouse events.)

I am in the midst of fixing this but it's a rather large patch, and I haven't completed all controls yet.  Once I do, it should be *much* clearer which control actually has focus, and I'll want to revisit tab order bugs then.

 TL;DR: I greatly appreciate the offer to help, but VB's tab order property just gets ignored anyway, so IMO let's not worry about it.  But I'd love to revisit when I have controls rendering properly and we can see which one(s) have focus at any given time, to make sure that the current implementation isn't buggy.

**Comment by jpbro** [comment]
No worries Tanner - I was just looking for some low hanging (and bland) fruit to grab onto when I saw the tab order stuff. Sounds like you have a much better solution in 7.0 already, so I'll move on :)

**Comment by tannerhelland** [comment]
With an 8.0 beta release now available, I'm content to close this issue for now.  I am happy to revisit specific papercuts as I encounter them (or they're reported to me), but I don't know how helpful a "master" papercut tracker like this is anymore.

Thank you again to @jpbro for his many contributions to this release, especially relating to improved tab key navigation as described above!
